iaso/utils/models/common.py

```diff
@@ -89,3 +90,47 @@ def check_instance_bulk_gps_push(queryset: QuerySet) -> (bool, Dict[str, List[in
         warnings["warning_overwrite"] = overwrite_ids
 
     return success, errors, warnings
+
+def check_instance_reference_bulk_link(queryset: QuerySet) -> (bool, Dict[str, List[int]], Dict[str, List[int]]): # type: ignore
+    from iaso.models.org_unit import OrgUnitReferenceInstance, OrgUnitType
+    queryset = queryset.annotate(
+        has_matching_reference_form=Exists(
+            OrgUnitType.reference_forms.through.objects.filter(
+                orgunittype_id=OuterRef("org_unit__org_unit_type"),  
+                form_id=OuterRef("form_id") 
+            )
+        )
+    )
+    
+    not_reference_instances = queryset.filter(has_matching_reference_form=False)
+    reference_instances = queryset.filter(has_matching_reference_form=True)
+
+    linked_and_unlinked = reference_instances.annotate(
+        is_linked=Exists(
+            OrgUnitReferenceInstance.objects.filter(instance_id=OuterRef("id"))
+        )
+    )
+
+    orgunit_form_pairs = list(linked_and_unlinked.values_list("org_unit_id", "form_id"))
+    duplicates_count = Counter(orgunit_form_pairs)
+    duplicate_org_unit_form_id_pairs = [pair for pair, count in duplicates_count.items() if count > 1]
+
+  
+    warnings = {}
+    errors = {}
+    infos = {}
+    success: bool = not duplicate_org_unit_form_id_pairs
+    if duplicate_org_unit_form_id_pairs:
+        errors["error_multiple_instances_same_org_unit"] = duplicate_org_unit_form_id_pairs
+        return success, infos, errors, warnings
+    
+    linked_instances = linked_and_unlinked.filter(is_linked=True)
+    unlinked_instances = linked_and_unlinked.filter(is_linked=False)   
+
+    if not_reference_instances:
+        warnings["no_reference_instances"] = [instance.id for instance in not_reference_instances]
+    infos["linked"] = [instance.id for instance in linked_instances]
+    infos["not_linked"] = [instance.id for instance in unlinked_instances]
```

Comment:
Since all you want is a list of IDs, we can avoid fetching every Instance field and speed up the search:

```suggestion
    linked_instances = linked_and_unlinked.filter(is_linked=True).values_list("id", flat=True)
    unlinked_instances = linked_and_unlinked.filter(is_linked=False).values_list("id", flat=True)   

    if not_reference_instances:
        warnings["no_reference_instances"] = not_reference_instances
    infos["linked"] = linked_instances
    infos["not_linked"] = unlinked_instances
```

Reply:
I forgot to include it here, but on line 106, there should also be a `.values_list("id", flat=True)` at the end of the line (otherwise you can't write my suggested line 131)